pkg/config/nodetreemodel/config.go

```diff
@@ -225,11 +225,63 @@ func (c *ntmConfig) SetDefault(key string, value interface{}) {
 	c.setDefault(key, value)
 }
 
+func (c *ntmConfig) findPreviousSourceNode(key string, source model.Source) (Node, error) {
+	previous := source.PreviousSource()
+	for {
+		if previous == model.SourceUnknown {
+			return nil, ErrNotFound
+		}
+		tree, err := c.getTreeBySource(previous)
+		if err != nil {
+			return nil, ErrNotFound
+		}
+		node := c.leafAtPathFromNode(key, tree)
+		if _, isMissing := node.(*missingLeafImpl); !isMissing {
+			return node, nil
+		}
+		previous = previous.PreviousSource()
+	}
+}
+
 // UnsetForSource unsets a config entry for a given source
-func (c *ntmConfig) UnsetForSource(_key string, _source model.Source) {
+func (c *ntmConfig) UnsetForSource(key string, source model.Source) {
 	c.Lock()
-	c.logErrorNotImplemented("UnsetForSource")
-	c.Unlock()
+	defer c.Unlock()
+
+	// Validate that the value at the given leaf is from the source we expect
+	var node Node = c.leafAtPathFromNode(key, c.root)
+	destNode, ok := node.(LeafNode)
+	if !ok {
+		return
+	}
+	if destNode.Source() != source {
+		return
+	}
+
+	// Find what the previous value used to be, based upon the previous source
+	prevNode, err := c.findPreviousSourceNode(key, source)
+	if err != nil {
+		return
+	}
+
+	// Get the parent node of the leaf we're unsetting
+	node = c.root
```

Comment:
We should also remove the node from the dedicated tree to avoid erroneous fallback and the unset value being show in the metadata.

I think it should go:
- Delete the leaf in the tree for the given source and delete inner nodes that don't have children anymore
- Delete with fallback from `root`

WDYT ?

Reply:
That sounds good, implemented it. Though my version doesn't clean up inner nodes that don't have children anymore. You think that's something we definitely need?

Reply:
I think it would be cleaner and keep the concept that innerNode can't be leaf node. Also, it might be confusing when dumped to YAML/JSON and sent has metadata. WDYT ?